chain/testing.go

```diff
@@ -74,64 +106,108 @@ func (f *Builder) BuildOn(parent *types.Block, build func(b *BlockBuilder)) *typ
 	if parent != nil {
 		tip = types.RequireNewTipSet(f.t, parent)
 	}
-	return f.BuildOnTip(tip, build)
+	return f.BuildOnTip(tip, 1, singleBuilder(build)).At(0)
 }
 
-// BuildManyOn builds a chain by invoking BuildOnTip `count` times.
-func (f *Builder) BuildManyOn(count int, parent types.TipSet, build func(b *BlockBuilder)) *types.Block {
-	require.True(f.t, count > 0, "")
-	for i := 0; i < count; i++ {
-		parent = types.RequireNewTipSet(f.t, f.BuildOnTip(parent, build))
+// BuildManyOn builds a chain by invoking BuildOnTip `height` times.
+func (f *Builder) BuildManyOn(height int, parent types.TipSet, build func(b *BlockBuilder)) types.TipSet {
+	require.True(f.t, height > 0, "")
+	for i := 0; i < height; i++ {
+		parent = f.BuildOnTip(parent, 1, singleBuilder(build))
 	}
-	return parent.At(0)
+	return parent
 }
 
-// BuildOnTip creates and returns a new block child of `parent`.
+// BuildOnTip creates and returns a new tipset child of `parent`.
+// The tipset carries `width` > 0 blocks with the same height and parents, but different tickets.
+// Note: the blocks will all have the same miner, which is unrealistic and forbidden by consensus;
+// generalise this to random miner addresses when that is rejected by the syncer.
 // The `build` function is invoked to modify the block before it is stored.
-func (f *Builder) BuildOnTip(parent types.TipSet, build func(b *BlockBuilder)) *types.Block {
-	ticket := make([]byte, binary.Size(f.seq))
-	binary.BigEndian.PutUint64(ticket, f.seq)
-	f.seq++
-
-	// Sum weight of parents' parent weight, plus one for each parent.
-	// Note: as with the state builder, we should probably factor this out.
-	parentWeight := types.Uint64(0)
-	for i := 0; i < parent.Len(); i++ {
-		parentWeight += parent.At(i).ParentWeight + 1
-	}
+func (f *Builder) BuildOnTip(parent types.TipSet, width int, build func(b *BlockBuilder, i int)) types.TipSet {
+	require.True(f.t, width > 0)
```

Comment:
baby nit: `require.NotZero` will give this a slightly better error message on failure (which is probably rare anyways)